man/man1/zhack.1

```diff
@@ -100,8 +100,9 @@ feature is now required to read the pool MOS.
 .Cm label repair
 .Ar device
 .Xc
-Repair corrupted labels by rewriting the checksum using the presumed valid
-contents of the label.
+Repair corrupted labels and recover detached devices
+by rewriting the checksum and uberblocks
+using the presumed valid contents of the label.
```

Comment:
We really have two cases here which we want to be careful not to conflate.

Case 1) The label has been damaged, or manually modified, such that the checksums are no longer valid.  The existing  command provides a way for someone knowledgeable about the on disk format to manually inspect the label and then regenerate the checksums if they believe the contents are reasonable.  This will allow it to be considered during an import.

Case 2) The device has been cleanly detached from the pool and the labels were overwritten to indicate this.  If I understand correctly, you'd like to ability to reconstruct a valid label using the uberblock stored in the first slot.

Let's keep these two cases distinct and instead add a new sub command for the undetach case.  How about something like `zpool repair undetach <device>`.

Reply:
I think you correctly describe the two cases.

To be clear, this should still remain within zhack (not the zpool command)? So, something like `zhack repair undetach`?

Otherwise, should be able to split the checksum repairing and detached device restoration into two separate functions. Since they both would write to different locations, it shouldn't matter which is called first (in the case that there is corruption on a detached device).

Reply:
What if the two commands were `zhack repair label` and `zhack repair undetach`? This makes what seems like a tidier subcommand structure, though it changes the existing command.

Reply:
I think I have a clearer image in my mind, now. Most of my first post was a bad take. The undetach operation alters the nvlist of a label, so it is necessary to recalculate the checksum for it. However, the label repair command could still exist in its original state to only fix checksums, without altering any data. By that logic, still shouldn't matter which is called first, though.

Reply:
Just finished a big refactor to implement the aforementioned commands. I'll probably have a fair amount of debugging to do. I decided to make `zhack repair undetach` see whether it's dealing with a label that already has correct checksums, before trying to reconstruct. `zhack repair label` will just try to fix the checksums.

Reply:
That sounds good.  I think we want to keep this functionality as part of `zhack`.  Using "undetach" seems like a pretty descripive name, but I don't feel strong about it, so if something better occurs to you while reworking this feel free to run with it.

Reply:
Alright, great. Agreed about zhack. I've been wondering if `zhack repair checksums` or something would be a better command name than `zhack repair label`, since both commands concern themselves with labels.

Reply:
Since `zhack` can operation on things other than vdev labels I'd suggest we keep the `label` subcommand.  How about something like this instead.  We could also assume `-c` option if no other options are specified for backwards compatibility.  Although since we're talking about `zhack` changing the command behavior wouldn't be a major concern.

```
zhack label repair [-cu] <device>
    -c repair corrupted label checksums
    -u restore the label on a detached device
```